src/Agent.php

```diff
@@ -721,11 +722,11 @@ public function requestAgent($endpoint): Response
                 $response = $httpClient->request('GET', $endpoint, []);
                 self::$found_address = $address;
                 break;
-            } catch (\GuzzleHttp\Exception\RequestException $exception) {
-                // got an error response, we don't need to try other addresses
-                break;
-            } catch (\Throwable $exception) {
-                // many addresses will be incorrect
+            } catch (\GuzzleHttp\Exception\ClientException $exception) {
+                // If a 4XX error seems to be coming from an agent, we can stop trying. For every other exception, we can try other addresses
+                if (str_contains($exception->getResponse()->getHeaderLine('Server'), 'libwww-perl-daemon')) {
```

Comment:
The initial purpose was to handle 405 responses. IMHO, we should only handle this specific case.
@g-bougard Do you agree ?

```suggestion
                // If a "405 Not allowed" error is returned by an agent expected address, we can stop trying. For every other exception, we can try other addresses.
                if ($exception->getResponse()->getStatusCode() === 405) {
                    // "405 Not allowed" status is returned when agent don't trust server
```

Reply:
A 405 doesn't seem like the right status code if the agent doesn't trust the server. This implies that the untrusted server could make the same request to the agent using a different method (GET, POST, etc) and the agent is expected to send an `Allow` header in the response indicating the allowed methods.

A 403 error is what should be used and seems like the only case where we know to stop trying to contact the agent at this time.

Reply:
Glpi-agent returns "403 Access denied" for untrusted ips. But this doesn't mean GLPI should not try other ips as routing can change the source ip to a trusted ip from glpi-agent point of view.

Maybe the best answer would be to try all ips and cache the first working ip to be reused first later.

Reply:
What are the performance implications? Are there cases where there are more than two different agent addresses or cases where GLPI contacts more than a single agent at a time (so not talking about clicking the buttons to request status or inventory in the computer form)?

Reply:
> Maybe the best answer would be to try all ips and cache the first working ip to be reused first later.

I'd agree. There is no way to really know why one of the address do not work, many are "guessed" at inventory time - and this also seem the simplest way to go (we just take care any of the request takes too long - perhaps adding a configuration parameter).
Caching should maybe help; but there is no way to know if the problem was temporary - so cache lifetime should  maybe not be too long.

I have absolutely no idea how often we really rely on that code part. From a core point of view it's only used on user interaction as far as I remember.

Reply:
It should also be used with glpi-inventory plugin with the "wake up agents" feature.

Reply:
It is pretty hard to know if an error result should be considered as temporary or not. The only thing we can improve here is to add a timeout to reduce the waiting time when the request is sent through a firewall that hangs it indefinitely, but it is out of the scope of the current PR.